core/src/main/java/io/undertow/server/handlers/URLDecodingHandler.java

```diff
@@ -55,38 +59,65 @@ public URLDecodingHandler(final HttpHandler next, final String charset) {
 
     @Override
     public void handleRequest(final HttpServerExchange exchange) throws Exception {
-        boolean decodeDone = exchange.getConnection().getUndertowOptions().get(UndertowOptions.DECODE_URL, true);
-        if (!decodeDone) {
-            final StringBuilder sb = new StringBuilder();
-            final boolean decodeSlash = exchange.getConnection().getUndertowOptions().get(UndertowOptions.ALLOW_ENCODED_SLASH, false);
-            exchange.setRequestPath(URLUtils.decode(exchange.getRequestPath(), charset, decodeSlash, false, sb));
-            exchange.setRelativePath(URLUtils.decode(exchange.getRelativePath(), charset, decodeSlash, false, sb));
-            exchange.setResolvedPath(URLUtils.decode(exchange.getResolvedPath(), charset, decodeSlash, false, sb));
-            if (!exchange.getQueryString().isEmpty()) {
-                final TreeMap<String, Deque<String>> newParams = new TreeMap<>();
-                for (Map.Entry<String, Deque<String>> param : exchange.getQueryParameters().entrySet()) {
-                    final Deque<String> newVales = new ArrayDeque<>(param.getValue().size());
-                    for (String val : param.getValue()) {
-                        newVales.add(URLUtils.decode(val, charset, true, true, sb));
-                    }
-                    newParams.put(URLUtils.decode(param.getKey(), charset, true, true, sb), newVales);
+        if (shouldDecode(exchange)) {
+            final StringBuilder sb = getStringBuilderForDecoding(exchange);
+            decodePath(exchange, charset, sb);
+            decodeQueryString(exchange, charset, sb);
+            decodePathTemplateMatch(exchange, charset, sb);
+        }
+        next.handleRequest(exchange);
+    }
+
+    // Returns true if the exchange should be decoded.  This method updates the ALREADY_DECODED
+    // attachment so that subsequent invocations will always return false.
+    private static boolean shouldDecode(final HttpServerExchange exchange) {
+        return !exchange.getConnection().getUndertowOptions().get(UndertowOptions.DECODE_URL, true)
+                && exchange.putAttachment(ALREADY_DECODED, Boolean.TRUE) == null;
```

Comment:
ALREADY_DECODED is only set or used if UndertowOptions.DECODE_URL has been set to `false`